### PR TITLE
chore(latest): get the latest version from the cdn

### PR DESCRIPTION
**Pull request by cnishina**
This reads the xml from the CDN to get the latest chromedriver, iedriver, or standalone version if the version is 'latest'. If the release is from Github, use the Github API to get the releases. Also store the downloaded information to a cache in the output directory (default: selenium/). If the file is older than one hour it will be rewritten.

When getting the status, we are no longer showing the default version. Default versions will be deprecated and will be removed from the config.json file.

### CODE REVIEW

**Comment by cnishina** [comment]
No, Craig, no. When doing the `start` command, should use the version associated with the 'latest' but not use the version called 'latest'

```
[20:15:48] I/start - java -Djava.security.egd=file:///dev/./urandom -Dwebdriver.chrome.driver=/home/ubuntu/webdriver-manager/selenium/chromedriver_latest -Dwebdriver.gecko.driver=/home/ubuntu/webdriver-manager/selenium/geckodriver-latest -jar /home/ubuntu/webdriver-manager/selenium/selenium-server-standalone-latest.jar -port 4444
```

**Comment by cnishina** [comment]
This is ready for review. Things that break due to getting the latest binary:

- e2e tests
- shutdown command `http://localhost:4444/selenium-server/driver/?cmd=shutDownSeleniumServer`

File: lib/binaries/gecko_driver_github.ts [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+  }
+
+  getVersionList(): Promise<string[]> {
+    return this.getJson().then(() => {
+      return this.getGeckoDriverList();
+    });
+  }
+
+  private getGeckoDriverList(): string[] {
+    this.versionsLookup = [];
```
Comment (sjelin) [inline lib/binaries/gecko_driver_github.ts]: Again, return this instead of setting a property.

File: lib/binaries/config_source.ts [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+    super();
+  }
+
+  protected getFileName(): string {
+    try {
+      fs.statSync(this.out_dir);
+    } catch (e) {
+      fs.mkdirSync(this.out_dir);
+    }
+    this.fileName = path.resolve(this.out_dir, this.name + '-response.xml');
```
Comment (sjelin) [inline lib/binaries/config_source.ts]: Why are you setting a property instead of just returning?  This makes the logic unnecessarily hard to follow.

File: lib/binaries/config_source.ts [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+    return this.fileName;
+  }
+
+  protected getXml(): Promise<any> {
+    this.getFileName();
+    let content = this.readResponse();
+    if (content != null) {
+      this.xml = content;
```
Comment (sjelin) [inline lib/binaries/config_source.ts]: Again, don't set a property, just return.  Side effects are bad and hard to follow

File: lib/binaries/config_source.ts [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+  opt_ignoreSSL: boolean;
+  opt_proxy: string;
+
+  abstract getUrl(version: string): Promise<{url: string, version: string}>;
+  abstract getVersionList(): Promise<string[]>;
+}
+
+export abstract class XmlConfigSource extends ConfigSource {
+  xml: any;
+  xmlUrl: string;
```
Comment (sjelin) [inline lib/binaries/config_source.ts]: Comment here that the concrete classes should set `xmlUrl` when initialized

File: lib/binaries/config_source.ts [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+    let retResult: any = null;
+    xml2js.parseString(xml, (err, result) => {
+      retResult = result;
+    });
+    return retResult;
+  }
+}
+
+export abstract class JsonConfigSource extends ConfigSource {
+  json: any;
```
Comment (sjelin) [inline lib/binaries/config_source.ts]: As with the `XmlConfigSource` comments, the `json` and `fileName` properties should be removed, and the `jsonUrl` property needs a comment

File: lib/binaries/iedriver_xml.ts [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
+      // Filter For IEDriverServer win 32. Removing option to download x64
+      if (contentKey.includes('IEDriverServer_Win32_')) {
+        versionPaths.push(contentKey);
+      }
+    }
+    return versionPaths;
+  }
+
+  private getLatestIEDriverVersion(): BinaryUrl {
```
Comment (sjelin) [inline lib/binaries/iedriver_xml.ts]: This is like the 4th time you've written a function for "get largest element in array".  Please write helper function:

```ts
getLatestVersion<T>(versionList: T[], comparison: (x: T, y: T) => boolean): T {
   ...
}
```

File: lib/binaries/iedriver_xml.ts [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+      } else if (semver.gt(version, latestVersion)) {
+        // Get the latest.
+        latestVersion = version;
+        latest = item;
+      }
+    }
+    return {url: latest, version: latestVersion};
+  }
+
+  private getSpecificIEDriverVersion(inputVersion: string): BinaryUrl {
```
Comment (sjelin) [inline lib/binaries/iedriver_xml.ts]: Same as above - write a helper function:

```ts
getSpecificVersion<T>(versionList: T[], isSpecificVersion: (version: T) => boolean): T {
   ...
}
```

File: lib/binaries/iedriver.ts [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+
+  prefix(): string {
+    return 'IEDriverServer';
+  }
+
+  suffix(): string {
+    return '.zip';
+  }
+
+  getUrl(version?: string): Promise<BinaryUrl> {
```
Comment (sjelin) [inline lib/binaries/iedriver.ts]: You write this exact function several times.  Please write a helper function, or make this the default implementation for `Binary`, or something

File: lib/binaries/standalone_xml.ts [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+  }
+
+  private getLatestStandaloneVersion(): BinaryUrl {
+    let list = this.getStandaloneList();
+    let standaloneVersion = null;
+    let latest = '';
+    let latestVersion = '';
+    for (let item of list) {
+      // Get a semantic version.
+      let version = item.split('selenium-server-standalone-')[1].replace('.jar', '');
```
Comment (sjelin) [inline lib/binaries/standalone_xml.ts]: As with `getLatestChromeDriverVersion` and `getSpecificChromeDriverVersion`, please move parsing/filtering into a helper function

File: lib/binaries/binary.ts [hunk 10]
```diff
@@ -9,103 +11,97 @@ export enum OS {
...
-  arch: string;
-
-  constructor(cdn?: string) {
-    this.cdn = cdn;
+  configSource: ConfigSource;
+
+  ostype: string = Config.osType();
+  osarch: string = Config.osArch();
+
+  alternativeDownloadUrl: string;  // override everything
```
Comment (cnishina) [inline lib/binaries/binary.ts]: Added a TODO comment

File: lib/binaries/binary.ts [hunk 11]
```diff
@@ -9,103 +11,97 @@ export enum OS {
...
+
+  cdn: string;      // The url host for XML reading or the base path to the url.
+  urlPath: string;  // The url path to download the binary. full url = baseUrl + downloadPath.
```
Comment (cnishina) [inline lib/binaries/binary.ts]: Done.

File: lib/binaries/binary.ts [hunk 12]
```diff
@@ -9,103 +11,97 @@ export enum OS {
...
+  opt_ignoreSSL: boolean;  // An optional ignore ssl.
+  opt_proxy: string        // An optional proxy.
+  downloadPath: string;    // The path to the file.
```
Comment (cnishina) [inline lib/binaries/binary.ts]: Done.

File: lib/binaries/binary.ts [hunk 13]
```diff
@@ -9,103 +11,97 @@ export enum OS {
...
+  downloadFile: string;    // The downloaded file name.
```
Comment (cnishina) [inline lib/binaries/binary.ts]: Done.

File: lib/binaries/config_source.ts [hunk 14]
```diff
@@ -0,0 +1,189 @@
+import * as fs from 'fs';
+import * as os from 'os';
+import * as path from 'path';
+import * as request from 'request';
+import * as semver from 'semver';
```
Comment (cnishina) [inline lib/binaries/config_source.ts]: Done.

File: lib/binaries/gecko_driver.ts [hunk 15]
```diff
@@ -1,9 +1,9 @@
-import * as path from 'path';
 import * as semver from 'semver';
```
Comment (cnishina) [inline lib/binaries/gecko_driver.ts]: Done.

File: lib/binaries/config_source.ts (hunk 4 above)
Comment (cnishina) [inline lib/binaries/config_source.ts]: Moving `xmlUrl` to the constructor

File: lib/binaries/config_source.ts (hunk 2 above)
Comment (cnishina) [inline lib/binaries/config_source.ts]: Agree... this is confusing. Done.

File: lib/cmds/start.ts [hunk 16]
```diff
@@ -114,124 +114,161 @@ function start(options: Options) {
...
-    args.push(
-        '-Dwebdriver.chrome.driver=' +
-        path.resolve(outputDir, binaries[ChromeDriver.id].executableFilename(osType)));
-    if (chromeLogs != null) {
-      args.push('-Dwebdriver.chrome.logfile=' + chromeLogs);
-    }
+    let chrome: ChromeDriver = binaries[ChromeDriver.id];
+    promises.push(
+        chrome.getUrl(chrome.versionCustom)
+            .then(binaryUrl => {
```
Comment (cnishina) [inline lib/cmds/start.ts]: Cleaning up. `binaryUrl` since it is not used.

File: lib/binaries/gecko_driver_github.ts (hunk 1 above)
Comment (cnishina) [inline lib/binaries/gecko_driver_github.ts]: modified to have a method return the version lookup list

File: lib/binaries/config_source.ts (hunk 3 above)
Comment (cnishina) [inline lib/binaries/config_source.ts]: Done.

File: lib/binaries/config_source.ts (hunk 5 above)
Comment (cnishina) [inline lib/binaries/config_source.ts]: Done.

File: lib/binaries/iedriver.ts (hunk 8 above)
Comment (cnishina) [inline lib/binaries/iedriver.ts]: implemented this at the binary level

File: lib/binaries/chrome_xml.ts [hunk 17]
```diff
@@ -0,0 +1,153 @@
...
+
+  /**
+   * Gets the latest item from the XML.
+   */
+  private getLatestChromeDriverVersion(): BinaryUrl {
+    let list = this.getChromeDriverList();
+    let chromedriverVersion = null;
+    let latest = '';
+    let latestVersion = '';
+    for (let item of list) {
```
Comment (cnishina) [inline lib/binaries/chrome_xml.ts]: removed method `getChromeDriverVersions`

File: lib/binaries/iedriver_xml.ts (hunk 6 above)
Comment (cnishina) [inline lib/binaries/iedriver_xml.ts]: I'd rather leave this as is. Getting the latest is different in chrome, selenium, ie, and gecko. To generalize this would make things more complicated for future work.

File: lib/binaries/iedriver_xml.ts (hunk 7 above)
Comment (cnishina) [inline lib/binaries/iedriver_xml.ts]: See above reasoning. Not a fan of making this generic.

File: lib/binaries/chrome_xml.ts [hunk 18]
```diff
@@ -0,0 +1,147 @@
...
+    return this.getXml().then(xml => {
+      let versionPaths: string[] = [];
+      let osType = this.getOsTypeName();
+
+      for (let content of xml.ListBucketResult.Contents) {
+        let contentKey: string = content.Key[0];
+
+        // Filter for 32-bit devices, make sure x64 is not an option
+        if (this.osarch !== 'x64' && contentKey.includes('64')) {
+          continue;
```
Comment (sjelin) [inline lib/binaries/chrome_xml.ts]: `continue` statements are bad style imo.  Please do:

```js
if (this.osarch === 'x64' || !contentKey.includes('64')) {
 ... // remainder of for loop
}
```

File: lib/binaries/chrome_xml.ts [hunk 19]
```diff
@@ -0,0 +1,147 @@
...
+      let chromedriverVersion = null;
+      let latest = '';
+      let latestVersion = '';
+      for (let item of list) {
+        // get a semantic version
+        let version = item.split('/')[0];
+        if (semver.valid(version) == null) {
+          version += '.0';
+          if (semver.valid(version) == null) {
+            continue;
```
Comment (sjelin) [inline lib/binaries/chrome_xml.ts]: Again, I'm not a fan of `continue`, but if this turns out to be a pain to remove you can leave it.

File: lib/binaries/chrome_xml.ts [hunk 20]
```diff
@@ -0,0 +1,147 @@
...
+      let itemFound = '';
+      let specificVersion = semver.valid(inputVersion) ? inputVersion : inputVersion + '.0';
+
+      for (let item of list) {
+        // Get a semantic version.
+        let version = item.split('/')[0];
+        if (semver.valid(version) == null) {
+          version += '.0';
+          if (semver.valid(version) == null) {
+            continue;
```
Comment (sjelin) [inline lib/binaries/chrome_xml.ts]: Same here.  Avoid `continue` if it can be done cleanly

File: lib/binaries/standalone_xml.ts [hunk 21]
```diff
@@ -0,0 +1,96 @@
...
+      let standaloneVersion = null;
+      let latest = '';
+      let latestVersion = '';
+      for (let item of list) {
+        // Get a semantic version.
+        let version = item.split('selenium-server-standalone-')[1].replace('.jar', '');
+
+        // Do not do beta versions for latest.
+        if (version.includes('beta')) {
+          continue;
```
Comment (sjelin) [inline lib/binaries/standalone_xml.ts]: `continue` is bad.  Please use:

```js
if (!version.includes('beta')) {
  ... // The rest of the for loop
}
```

File: lib/binaries/chrome_xml.ts (hunk 20 above)
Comment (sjelin) [inline lib/binaries/chrome_xml.ts]: Same here. Avoid `continue` if it can be done cleanly

File: lib/binaries/standalone_xml.ts (hunk 21 above)
Comment (cnishina) [inline lib/binaries/standalone_xml.ts]: Done.

File: lib/binaries/chrome_xml.ts (hunk 18 above)
Comment (cnishina) [inline lib/binaries/chrome_xml.ts]: Done.

File: lib/binaries/chrome_xml.ts (hunk 19 above)
Comment (cnishina) [inline lib/binaries/chrome_xml.ts]: Done.

File: lib/binaries/chrome_xml.ts (hunk 20 above)
Comment (cnishina) [inline lib/binaries/chrome_xml.ts]: Done.
